src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,1723 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
```

Comment:
>namespace [](start = 0, length = 9)

I don't see any tests comparing nullable tuples of different types, e.g. `(int, int)? t1` and `(long, long)? t2`, where conversions are needed to do the tests. Can you please add such tests? Also, it would be nice to have at least one test involving a user-defined conversion in the mix with nullable too.

Reply:
Good point.
Some other test ideas: 
- nullable tuples used with `!=` (to make sure the binary logic around HasValue is correct).

---
In reply to: [153580386](https://github.com/dotnet/roslyn/pull/23356#discussion_r153580386) [](ancestors = 153580386)

Reply:
Also, please add a test with tuple cardinality 1 (using a long tuple's Rest field).

---
In reply to: [153588129](https://github.com/dotnet/roslyn/pull/23356#discussion_r153588129) [](ancestors = 153588129,153580386)

Reply:
Are there tests for tuples passed as `in`?